core-bundle/tests/Filesystem/PermissionCheckingVirtualFilesystemTest.php

```diff
@@ -110,5 +170,90 @@ public static function provideOperationsThatShouldBeDenied(): iterable
             ContaoCorePermissions::USER_CAN_UPLOAD_FILES,
             'Access denied to upload files at location "bar".',
         ];
+
+        yield 'get' => [
+            'get',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'listContents' => [
+            'listContents',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'getLastModified' => [
+            'getLastModified',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'getFileSize' => [
+            'getFileSize',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'getMimeType' => [
+            'getMimeType',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'getExtraMetadata' => [
+            'getMimeType',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'setExtraMetadata' => [
+            'setExtraMetadata',
+            ['foo', []],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+
+        yield 'generatePublicUri' => [
+            'generatePublicUri',
+            ['foo'],
+            [ContaoCorePermissions::USER_CAN_ACCESS_PATH, 'files/foo'],
+            'Access denied to access path at location "foo".',
+        ];
+    }
+
+    /**
+     * @dataProvider provideInvalidPaths
+     */
+    public function testDisallowsAccessForInvalidPaths(string $invalidPath, string $expectedMessage): void
+    {
+        $permissionCheckingVirtualFilesystem = new PermissionCheckingVirtualFilesystem(
+            $this->createMock(VirtualFilesystem::class),
+            $this->createMock(Security::class),
+        );
+
+        $this->expectException(AccessDeniedException::class);
+        $this->expectExceptionMessage($expectedMessage);
+
+        $permissionCheckingVirtualFilesystem->has($invalidPath);
+    }
+
+    public static function provideInvalidPaths(): iterable
+    {
+        yield 'relative path escaping boundary' => [
+            '../foo',
```

Comment:
I would like to have one on top that uses `./../` as that would not start with `..` - should be covered by `Path::canonicalize()` but let's be sure :)

Reply:
Added in 0c7b2387e49b0c8015cc716d04b2eacf19aa72fc.